ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java

```diff
@@ -88,6 +88,15 @@ default CompletableFuture<?> write(LogEntryProto entry) {
       return CompletableFuture.completedFuture(null);
     }
 
+    /**
+     * Write asynchronously the state machine data in the given log entry to this state machine.
+     *
+     * @return a future for the write task
+     */
+    default CompletableFuture<?> write(LogEntryProto entry, TransactionContext context) {
```

Comment:
Shall we add a `@Nullable` annotation here to notify user the `context` will be null when the peer is a follower?

Reply:
That's a good point!  Follower currently creates context in `RaftServerImpl.applyLogToStateMachine`.  We should create a context instead of passing null and then the context can be reused.  Let me think about it